docs/design/resource-qos.md

```diff
@@ -0,0 +1,254 @@
+
+## Introduction
+
+This document describes the way Kubernetes provides different levels of Quality of Service to pods depending on what they *request*.
+Pods that need to stay up reliably can request guaranteed resources, while pods with less stringent requirements can use resources with weaker or no guarantee.
+
+Specifically, for each resource, containers specify a request, which is the amount of that resource that the system will guarantee to the container, and a limit which is the maximum amount that the system will allow the container to use.
+The system computes pod level requests and limits by summing up per-resource requests and limits across all containers.
+When request == limit, the resources are guranteed, and when request < limit, the pod is guaranteed the request but can opportunistically scavenge the difference between request and limit if they are not being used by other containers.
+This allows Kubernetes to oversubscribe nodes, which increases utilization, while at the same time maintaining resource guarantees for the containers that need guarantees.
+Borg increased utilization by about 20% when it started allowing use of such non-guaranteed resources, and we hope to see similar improvements in Kubernetes.
+
+## Requests and Limits
+
+For each resource, containers can specify a resource request and limit, 0 <= request <= limit <= Infinity.
+If a pod is successfully scheduled, the container is guaranteed the amount of resources requested.
+Scheduling is based on `requests` and not `limits`.
+The pods and its containers will not be allowed to exceed the specified limit.
+How the request and limit are enforced depends on whether the resource is [compressible or incompressible](resources.md).
+
+### Compressible Resource Guarantees
+
+- For now, we are only supporting CPU.
+- Pods are guaranteed to get the amount of CPU they request, they may or may not get additional CPU time (depending on the other jobs running).
+- Excess CPU resources will be distributed based on the amount of CPU requested. For example, suppose container A requests for 600 milli CPUs, and container B requests for 300 milli CPUs. Suppose that both containers are trying to use as much CPU as they can. Then the extra 10 milli CPUs will be distributed to A and B in a 2:1 ratio (implementation discussed in later sections).
+- Pods will be throttled if they exceed their limit. If limit is unspecified, then the pods can use excess CPU when available.
+
+### Incompressible Resource Guarantees
+
+- For now, we are only supporting memory.
+- Pods will get the amount of memory they request, if they exceed their memory request, they could be killed (if some other pod needs memory), but if pods consume less memory than requested, they will not be killed (except in cases where system tasks or daemons need more memory).
+- Pods will be killed if they use more memory than their limit.
```

Comment:
we should clarify the meaning of killed.  people often seem to confuse killed with pod moves to a terminal state, when in this situation it more than likely means processes are restarted.